File: OS Lab/Lab/dining.c
#include<stdio.h>
#include<semaphore.h>
#include<pthread.h>

#define N 5
#define THINKING 0
#define HUNGRY 1
#define EATING 2
#define LEFT (ph_num+4)%N
#define RIGHT (ph_num+1)%N
 
sem_t mutex;
sem_t S[N];
 
int ate[N];
int thought[N];
 	
void *philospher(void *num);
void take_fork(int);
void think(int);
void test(int);
 
int state[N];
int phil_num[N]={0,1,2,3,4};
 
int main()
{
    int i;
    pthread_t thread_id[N];
    sem_init(&mutex,0,1);
    printf("OUTPUT\n");
    for(i=0;i<N;i++){
        sem_init(&S[i],0,0);
        ate[i] = 0;
        thought[i] = 0;
       }
    for(i=0;i<N;i++)
    {
        pthread_create(&thread_id[i],NULL,philospher,&phil_num[i]);
        printf("Philosopher %d is thinking\n",i+1);
    }
    for(i=0;i<N;i++)
        pthread_join(thread_id[i],NULL);
}
 
void *philospher(void *num)
{
    while(1)
    {
    int *i = num;
    	if (ate[*i] && thought[*i]){
    		printf("*Philosopher %d left the table\n", (*i)+1);
    		return;
    	}
        
        sleep(1);
        take_fork(*i);
        sleep(0);
        think(*i);
    }
}
 
void take_fork(int ph_num)
{
	if(ate[ph_num])
		return;
    sem_wait(&mutex);
    state[ph_num] = HUNGRY;
    test(ph_num);
    sem_post(&mutex);
    sem_wait(&S[ph_num]);
    sleep(1);
}
 
void test(int ph_num)
{
    if (state[ph_num] == HUNGRY && state[LEFT] != EATING && state[RIGHT] != EATING)
    {
        state[ph_num] = EATING;
        sleep(2);
        printf("Philosopher %d takes fork %d and %d\n",ph_num+1,LEFT+1,ph_num+1);
        printf("Philosopher %d is Eating\n",ph_num+1);
        ate[ph_num] = 1;
        sem_post(&S[ph_num]);
    }
}
 
void think(int ph_num)
{
	if(thought[ph_num])
		return;
		
    sem_wait(&mutex);
    state[ph_num] = THINKING;
    printf("Philosopher %d puts fork %d and %d down\n",ph_num+1,LEFT+1,ph_num+1);
    printf("Philosopher %d is thinking\n",ph_num+1);
    thought[ph_num] = 1;
    test(LEFT);
    test(RIGHT);
    sem_post(&mutex);
}	

File: OS Lab/Lab/dining.java
class fork
{
	boolean available=true;
	int num;
	fork(int n)
	{
	num=n;
	}
	
	synchronized  void take()
	{    
		String s=Thread.currentThread().getName();
		try
		{
		       if(available==false)
		          {
		 		System.out.println(s+" is thinking. . . .");
		 		wait();
			  }
		}
		catch(Exception e)
		{
		}
		System.out.println(s+" took fork "+num);
	    	available = false;
	}
	
	synchronized  void  putfork()
	{
		String s=Thread.currentThread().getName();
		System.out.println(s+" put fork "+num+" down");
		 available=true;
		 notify();
	}	 
}

class phil implements Runnable
{	
	int num;
	fork left,right;
	Thread t;
	phil(int n,String s,fork l,fork r)
	{
		left=l;
		right=r;
		num=n;
		t=new Thread(this,s);
		t.start();
	}
	
	public void run()
	{
		String s=Thread.currentThread().getName();
		if(num%2==0)
		{  
			right.take();
			left.take();
			try
			{
				System.out.println(s+" is eating. . . .");
				Thread.sleep(1000);
			}
			catch(Exception e)
			{
			}
			left.putfork();
			right.putfork();

		}
		else
		{
		        left.take();
			right.take();
			try
			{
				System.out.println(s+" is eating. . . .");
				Thread.sleep(1000);
			}
			catch(Exception e)
			{
			}
			right.putfork();
			left.putfork();
		}
	}

		
}


class dining
{
	public static void main(String[] args)
	{       
		fork f1=new fork(1);
		fork f2=new fork(2);
		fork f3=new fork(3);
		fork f4=new fork(4);
		fork f5=new fork(5);
		 
		phil p1=new phil(1,"philosopher 1",f1,f2);
		phil p2=new phil(2,"philosopher 2",f2,f3);
		phil p3=new phil(3,"philosopher 3",f3,f4);
		phil p4=new phil(4,"philosopher 4",f4,f5);
		phil p5=new phil(5,"philosopher 5",f5,f1);
				
	}
}
